modular_skyrat/modules/imenu/code/imenu.dm

```diff
@@ -0,0 +1,268 @@
+	for(var/requirement in interaction_requires)
+		switch(requirement)
+			if(INTERACTION_REQUIRE_SELF_HAND)
+				if(!user.get_active_hand())
+					return FALSE
+			if(INTERACTION_REQUIRE_SELF_SPEAK)
+				if(!user.can_speak())
+					return FALSE
+			if(INTERACTION_REQUIRE_TARGET_HAND)
+				if(!target.get_active_hand())
+					return FALSE
+			if(INTERACTION_REQUIRE_TARGET_SPEAK)
+				if(!target.can_speak())
+					return FALSE
+			else
+				message_admins("Unimplemented interaction requirement '[requirement]'. Blame coders.")
+	return TRUE
+
+/datum/interaction/proc/act(mob/living/user, mob/living/target)
+	if(!message)
+		message_admins("Interaction had a null message list. '[name]'")
+	if(!islist(message) && istext(message))
+		message_admins("Deprecated message handling for '[name]'. Correct format is a list with one entry. This message will only show once.")
+		message = list(message)
+	var/msg = pick(message)
+	// We replace %USER% with nothing because manual_emote already prepends it.
+	msg = trim(replacetext(replacetext(msg, "%TARGET%", "[target]"), "%USER%", ""), INTERACTION_MAX_CHAR)
+	user.manual_emote(msg)
+	if(sound_use)
+		if(isnull(sound_possible))
+			message_admins("Interaction has sound_use set to TRUE but does not set sound! '[name]'")
+			return
+		else if(islist(sound_possible))
+			sound_cache = sound(pick(sound_possible))
+		else sound_cache = sound(sound_possible)
+		for(var/mob/mob in view(sound_range, user))
+			SEND_SOUND(sound_cache, mob)
+
+/datum/component/interactable
+	var/mob/self = null
+	var/interact_last = 0
+	var/interact_next = 0
+
+/datum/component/interactable/proc/can_interact(datum/interaction/interaction, mob/user)
+	if(!interaction.allow_act(user, self))
+		return FALSE
+	if(!interaction.distance_allowed && !user.Adjacent(self))
+		return FALSE
+	if(interaction.category == INTERACTION_CAT_HIDE)
+		return FALSE
+	if(self == user && interaction.usage == INTERACTION_OTHER)
+		return FALSE
+	return TRUE
+
+/datum/component/interactable/proc/make_interact_list(datum/component/interactable/other)
+	populate_interaction_instances()
+	var/list/ints = list()
+	for(var/interaction in GLOB.interaction_instances)
+		if(other.can_interact(GLOB.interaction_instances[interaction], self))
+			ints += GLOB.interaction_instances[interaction]
+	return ints
+
+/datum/component/interactable/proc/mil_mob(mob/user)
+	var/datum/other = user.GetComponent(/datum/component/interactable)
+	if(!other)
+		CRASH("Unable to locate the interactable component for the given mob.")
+	return make_interact_list(other)
+
+/datum/component/interactable/ui_interact(mob/user, datum/tgui/ui)
+	ui = SStgui.try_update_ui(user, src, ui)
+	if(!ui)
+		ui = new(user, src, "InteractionMenu")
+		ui.set_autoupdate(TRUE)
+		ui.open()
+
+/datum/component/interactable/ui_status(mob/user, datum/ui_state/state)
+	return UI_INTERACTIVE // This UI is always interactive as we handle distance flags via can_interact
+
+/mob/ComponentInitialize()
```

Comment:
Pretty sure TG is moving away from ComponentInitialize() in favor of doing that in Initialize()
Not a criticism, just something to keep in mind

Reply:
Also, ghosts and newplayers having interactions?

Reply:
Or even /carbon so you don't have simplemob.. "interactions"